### ISSUE TITLE
How to disable media Queue button in Android Auto when playing live stream.

### ISSUE THREAD

**Issue by Guardiannw** [androidx/media#339]
Question:

Under the previous version of the media library, when playing a live stream (HLS) for a radio station, the Queue button did not appear in Android Auto when playing a media item.  However, after migrating to the new Media3 library, the Queue seems to be there by default.  I have tried to look through the documentation, but cannot find a way to easily disable the Queue button on Android Auto.  I assume it has something to do with the new implementation because it was not there for us on the old implementation.  Is there any way that I can easily disable the viewing of the queue in Android Auto via the library.

**Comment by marcbaechinger** [comment]
I assume that in the legacy world the queue button is shown when [`FLAG_HANDLES_QUEUE_COMMANDS`](https://developer.android.com/reference/android/support/v4/media/session/MediaSessionCompat#FLAG_HANDLES_QUEUE_COMMANDS()) was set in `MediaSessionCompat`. 

Media3 [maps `COMMAND_CHANGE_MEDIA_ITEMS` to the flags](https://github.com/androidx/media/blob/release/libraries/session/src/main/java/androidx/media3/session/MediaSessionLegacyStub.java#L329) to be set to the session.

I think the only thing you can do, is to use a `ForwardingPlayer` and override the methods that get the available commands:

```kotlin
    val forwardingPlayer =
      object : ForwardingPlayer(player) {

        override fun isCommandAvailable(command: Int): Boolean {
          if (command == COMMAND_CHANGE_MEDIA_ITEMS) {
            return false;
          }
          return super.isCommandAvailable(command)
        }

        override fun getAvailableCommands(): Player.Commands {
          val commands = super.getAvailableCommands()
          if (commands.contains(COMMAND_CHANGE_MEDIA_ITEMS)) {
            return commands.buildUpon().remove(COMMAND_CHANGE_MEDIA_ITEMS).build()
          }
          return commands;
        }
      }
    
    MediaSession.Builder(context, forwardingPlayer).build()
    ...
```

That should work but is not really nice, because with this all the other controllers (including Media3 controllers) can't use that `COMMAND_CHANGE_MEDIA_ITEMS` either because it is removed from the player. I don't think there is another solution with the current version though. 

It may be sufficient for you when you do a 1:1 migration to Media3, because you didn't set `FLAG_HANDLES_QUEUE_COMMANDS` in the legacy world (or you didn't use a `QueueEditor` in case you were using `MediaSessionConnector`).  In case your own app would want to use a Media3 controller internally, this could be probably problematic.

However, can you as a first step test whether this actually works the way I assume above?

**Comment by Guardiannw** [comment]
Thanks @marcbaechinger for responding.  I have just tried your suggestions, but unfortunately, the queue button is still there.  I have also tried removing these commands in the `onConnect` method of the `MediaLibrarySession.Callback`, but the issue still persists.

**Comment by marcbaechinger** [comment]
Ok, thanks! Will investigate.

**Comment by marcbaechinger** [comment]
The criteria is whether the queue is empty or not. 

With Media3, the queue of the legacy session is a function of the media items added to the player. So if there is a media item playing, then the queue is never empty.

This is a regression compared to what apps can do with the legacy `MediaSessionCompat` API. I think we need to fix this. because this

I changed this issue to a bug and will update this issue when we have news.

**Comment by DenysLoboda** [comment]
@marcbaechinger do we have some option to override `MediaSession::onTimelineChanged`? Or perhaps there is some other workaround?

**Comment by DenysLoboda** [comment]
In fact, when using the support library, this behavior is also observed. I managed to get the playlist off by overriding `MediaMetadataProvider` and assuming `QueueNavigator` is not set. Are there any equivalents for this for `media3`?

**Comment by marcbaechinger** [comment]
With the commit above that landed in the main branch and will be released with 1.2.0, an app can now remove the queue button in Android Auto/Automotive OS by removing `COMMNAD_GET_TIMELINE`:

```kotlin
override fun onConnect(
  session: MediaSession,
  controller: MediaSession.ControllerInfo
): MediaSession.ConnectionResult {
  if (session.isMediaNotificationController(controller)) {
    val playerCommands =
      Player.Commands.Builder().addAllCommands().remove(COMMAND_GET_TIMELINE).build()
    return MediaSession.ConnectionResult.AcceptedResultBuilder(session)
      .setAvailablePlayerCommands(playerCommands)
      .build()
  }
}
```

With this the queue is not set on the framework session and the button is not displayed y Auto. 

An app can change this setting during runtime like for instance when the user adds a playlist instead a radio stream and an app want to alternate between the two behaviors:

```kotlin
session.mediaNotificationControllerInfo?.let {
  val playerCommands = Player.Commands.Builder().addAllCommands()
  val sessionCommands = getSessionCommandsIncludingCustomCommands() // provided by app
  session.setAvailableCommands(it, sessionCommands, playerCommands)
}
```

An app can do this even when the app wants to support the System UI notification in parallel. `COMMAND_GET_TIMELINE` is not required for supporting the System UI media notification.

**Comment by marcbaechinger** [comment]
@Guardiannw Sorry for the wait and thanks for your report!

@DenysLoboda 

> when using the support library, this behavior is also observed. 
> Are there any equivalents for this for media3?

You can remove `COMMAND_GET_TIMELINE` as explained above.

> do we have some option to override MediaSession::onTimelineChanged? 

Na, there is no such method. I think we the option added now, you can now do what you want by API.

I'm closing as the question is answered and we now have an API for this. Please open a new issue in case you find issues with the provided solution.
